### PR TITLE
no strings

### PR DESCRIPTION
**Pull request by bananaturtlesandwich**
this pr essentially minimises a lot of the string creation in terms of FName content and getting from name map
the main change for this is making the `get_content` signature into `fn get_content<T>(&self, func: impl FnOnce(&str) -> T) -> T` and similarly for `get_name_reference`
Using a closure means that the lifetime of the name ref is valid which means you don't need to make a String
Of course you can still use the old behaviour with `get_content_owned` and similarly `get_name_reference_owned`
for some convenience the `is` and `ends_with` methods have been implemented for fnames since it's very common to check these things as shown in the amount of times they've already been used here
Unfortunately I couldn't figure out a way to cut the string creation in `asset/mod.rs` line 200 but other than that pretty much everything involving string creation from names where a string wasn't being stored somewhere has been optimised :)

### CODE REVIEW

File: unreal_asset/src/asset/name_map.rs [hunk 1]
```diff
@@ -79,14 +79,16 @@ impl NameMap {
     }
 
     /// Get a name reference by an FName map index
-    pub fn get_name_reference(&self, index: i32) -> String {
+    pub fn get_name_reference(&self, index: i32) -> &str {
+        // to avoid the panic this could return an option instead
+
         if index < 0 {
-            return (-index).to_string(); // is this right even?
+            return "too low";
```
Comment (localcc) [inline unreal_asset/src/asset/name_map.rs:86]: maybe make it more descriptive like "ERROR: name reference index too low"?

Comment (bananaturtlesandwich) [reply unreal_asset/src/asset/name_map.rs:86]: sure can do

**Review by localcc — APPROVED** [review]

**Comment by bananaturtlesandwich** [comment]
not sure what I should do for the merge conflict since idk why it's lazy_static now - should I just use new one?

**Comment by localcc** [comment]
yeah that works

**Comment by localcc** [comment]
just make sure the list content is the same so there's no regressions

File: unreal_asset/src/types/fname.rs [hunk 2]
```diff
@@ -106,21 +106,35 @@ impl FName {
...
     }
 
-    /// Get this FName instance number
+    /// Get this `FName`'s content as a `String`
+    pub fn get_owned_content(&self) -> String {
+        self.get_content(str::to_string)
+    }
+
+    /// Checks if an `FName`'s content is the given `&str`
+    pub fn is(&self, pat: impl AsRef<str>) -> bool {
```
Comment (konsti219) [inline unreal_asset/src/types/fname.rs:128]: how about doing this via `ops::Eq`?

**Review by konsti219 — CHANGES_REQUESTED** [review]

File: unreal_asset/src/types/fname.rs (hunk 2 above)
Comment (bananaturtlesandwich) [reply unreal_asset/src/types/fname.rs:128]: ooh that's a good idea

Comment (bananaturtlesandwich) [reply unreal_asset/src/types/fname.rs:128]: wait where is that? https://doc.rust-lang.org/std/ops/index.html

Comment (konsti219) [reply unreal_asset/src/types/fname.rs:128]: oh, == is actually std::cmp

Comment (bananaturtlesandwich) [reply unreal_asset/src/types/fname.rs:128]: 👌

**Comment by bananaturtlesandwich** [comment]
@localcc looks like the lazy_static! version isn't what the crate's geared up for so I'll convert back and see how it goes

**Review by konsti219 — APPROVED** [review]

File: unreal_asset/src/asset/name_map.rs [hunk 3]
```diff
@@ -91,6 +91,11 @@ impl NameMap {
         &self.name_map_index_list[index as usize]
     }
 
+    /// Get a name reference by an FName map index as a `String`
+    pub fn get_owned_name_reference(&self, index: i32) -> String {
```
Comment (konsti219) [inline unreal_asset/src/asset/name_map.rs:95]: There is not really an owned reference. This could just be called `get_owned_name`.

Comment (bananaturtlesandwich) [reply unreal_asset/src/asset/name_map.rs:95]: lol true

**Review by konsti219 — APPROVED** [review]

**Pull request merged**